### PR TITLE
Implement individual immigration

### PR DESCRIPTION
**Pull request by zmbc**
### Description

- *Category*: implementation
- *JIRA issue*: [MIC-3777](https://jira.ihme.washington.edu/browse/MIC-3777)
- *Research reference*: https://vivarium-research.readthedocs.io/en/latest/models/concept_models/vivarium_census_synthdata/concept_model.html#id9

### Changes and notes

As noted in the code comments, this will get cleaner when we get `state` and `puma` out of the population table.

**Depends on #196, review that first!**

### Verification and Testing

Integration tests passing. I also ran them with the number of immigrants artificially set to 100 per timestep of each type to really stress test it.

I ran the simulation to completion to check that observers work.

### CODE REVIEW

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 1]
```diff
@@ -289,6 +297,102 @@ def initialize_newborns(self, pop_data: SimulantData) -> None:
...
+        ).astype(existing_simulants["household_id"].dtype)
+
+        # We avoid non-reference-person immigrants having these relationships, because
+        # otherwise they may cause there to be an impossible number of said relationships
+        # within a household, i.e. >2 parents or >1 spouse/partner
+        new_simulants["relation_to_household_head"] = (
+            new_simulants["relation_to_household_head"]
+            .replace(
+                {
+                    "Parent": "Other relative",
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/population.py:375]: What about father, mother, step-father, etc.? There are more relationships that imply uniqueness than just these. Or We could add a third grandmother, which is also impossible.

**Review by rmudambi — APPROVED** [review]

File: src/vivarium_census_prl_synth_pop/components/population.py (hunk 1 above)
Comment (zmbc) [reply src/vivarium_census_prl_synth_pop/components/population.py:375]: Mother and father are both "parent", which is already listed here. I think I should add parent-in-law to this list, and to the documentation.

There is no step-father relationship; that would be "other relative" already. The same is true of grandparent. The full list of relationships is here: https://github.com/ihmeuw/vivarium_census_prl_synth_pop/blob/3f7cf37eef13608d6dd9e246bcce6e4a201cb43f/src/vivarium_census_prl_synth_pop/constants/metadata.py#L150-L170

File: integration_tests/test_immigration.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    for before, after in zip(populations, populations[1:]):
+        new_simulants = after.index.difference(before.index)
+        immigrants = after.index[after["parent_id"] == -1].intersection(new_simulants)
+        timestep_values.append((before, after, after.loc[immigrants]))
+
+    return timestep_values
+
+
+def test_there_is_immigration(immigrants_by_timestep):
+    # This is a very rare event, so we can't assert that it happens
```
[the remark this answers is not in the document]
Comment (zmbc) [reply integration_tests/test_immigration.py:20]: Yes, I'll set some conservative bounds. Only GQ immigration is so rare that no meaningful lower bound is possible.

File: src/vivarium_census_prl_synth_pop/components/immigration.py [hunk 3]
```diff
@@ -85,3 +145,45 @@ def _immigrants_per_time_step(self, immigrants, configuration):
...
+    def _round_stochastically(self, num, additional_key):
+        """
+        Implements stochastic rounding.
+        For example, a value of 2.23 will round up to 3 with 23% probability, and down to 2
+        the rest of the time.
+        If we used traditional rounding, a value of e.g. 0.4 would round down to 0 on every
+        single timestep, and that immigration event would *never* happen.
+        """
+        return floor(
+            num + self.randomness.get_draw(pd.Index([0]), additional_key=additional_key)
```
[the remark this answers is not in the document]
Comment (zmbc) [reply src/vivarium_census_prl_synth_pop/components/immigration.py]: Here's what I came up with: https://github.com/ihmeuw/vivarium_census_prl_synth_pop/blob/eaeb9e17f61034e1d8368b2121ab1ddc8a3d0d30/src/vivarium_census_prl_synth_pop/components/immigration.py#L190-L199

[the remark this answers is not in the document]
Comment (rmudambi) [reply src/vivarium_census_prl_synth_pop/components/immigration.py]: I think this is the clearest of the three.

**Review by rmudambi — APPROVED** [review]

**Pull request merged**